---
title: "Sims_microbial_load"
output:
  html_document: default
  pdf_document: default
date: "2025-05-16"
editor_options:
  chunk_output_type: console
---

## Mars 2020 
```{r, warning=FALSE}
library(strainspy)
se_98_abundance <- readRDS("TEST_DATA/mars_2020/Filtered/mars_2020_profile_gtdb_220_id98_abundance.rds")
se_metaphlan <- readRDS("TEST_DATA/mars_2020/Filtered/mars_2020_metaphlan_profile.rds")
se_98_ANI <- readRDS("TEST_DATA/mars_2020/Filtered/mars_2020_profile_gtdb_220_id98_ANI.rds")
load_galaxy = readRDS("TEST_DATA/mars_2020/Filtered/microbial_load_galaxy.rds")
load_metacardis = readRDS("TEST_DATA/mars_2020/Filtered/microbial_load_metacardis.rds")
metadata = readRDS("TEST_DATA/mars_2020/Filtered/metadata.rds")

## add load data
if(all(names(load_galaxy) == metadata$Accession) ) metadata$load_g = load_galaxy else cat("Names not in order")
if(all(names(load_metacardis) == metadata$Accession) ) metadata$load_m = load_metacardis else cat("Names not in order")

se_metaphlan = strainspy:::modify_metadata(se = se_metaphlan, meta_data = metadata)
se_98_abundance = strainspy:::modify_metadata(se = se_98_abundance, meta_data = metadata)
se_98_ANI = strainspy:::modify_metadata(se = se_98_ANI, meta_data = metadata)

se_98_ANI <- filter_by_presence(se_98_ANI, min_nonzero = 10)
se_98_abundance <- filter_by_presence(se_98_abundance, min_nonzero = 5)
se_metaphlan <- filter_by_presence(se_metaphlan, min_nonzero = 5)

design <- as.formula("~ cohort + timepoint + age + BMI + gender + antibiotics + probiotics + yogurt + pain_meds + (1 + timepoint | patientID)")

fit_ANI = glmFit(se_98_ANI, design, 10L)
top_hits(fit_ANI, coef = 2)
top_hits(fit_ANI, coef = 3)
plot_manhattan(fit_ANI, coef = 2)
plot_manhattan(fit_ANI, coef = 3)
contigs = unique(c(top_hits(fit_ANI, 2)$Contig_name, top_hits(fit_ANI,3)$Contig_name)); if(length(contigs) > 0) plot_violin(se_98_abundance, 'cohort', contigs, drop_zeros = T, contig_names = NULL)


design <- as.formula("~ cohort + timepoint + age + BMI + gender + antibiotics + probiotics + yogurt + pain_meds + (1 + timepoint | patientID)")

fit_ANI_zb = glmZiBFit(se_98_ANI, design, 10L)
top_hits(fit_ANI, coef = 2)
top_hits(fit_ANI, coef = 3)
plot_manhattan(fit_ANI, coef = 2)
plot_manhattan(fit_ANI, coef = 3)
contigs = unique(c(top_hits(fit_ANI, 2)$Contig_name, top_hits(fit_ANI,3)$Contig_name)); if(length(contigs) > 0) plot_violin(se_98_abundance, 'cohort', contigs, drop_zeros = T, contig_names = NULL)


design <- as.formula("~ cohort + (1 + timepoint | patientID)")

fit_AB_skani = abundanceFit(se = se_98_abundance, design = design, nthreads = 10L, transform = 'CLR')
top_hits(fit_AB_skani, coef = 2)
top_hits(fit_AB_skani, coef = 3)
plot_manhattan(fit_AB_skani)
contigs = unique(c(top_hits(fit_AB_skani, 2)$Contig_name, top_hits(fit_AB_skani,3)$Contig_name)); if(length(contigs) > 0) plot_violin(se_98_abundance, 'cohort', contigs, drop_zeros = T, contig_names = NULL)

fit_AB_metaphlan = abundanceFit(se = se_metaphlan, design = design, nthreads = 10L, transform = 'CLR')
top_hits(fit_AB_metaphlan, coef = 2)
top_hits(fit_AB_metaphlan, coef = 3)
plot_manhattan(fit_AB_metaphlan)
contigs = unique(c(top_hits(fit_AB_metaphlan, 2)$Contig_name, top_hits(fit_AB_metaphlan,3)$Contig_name)); if(length(contigs) > 0) plot_violin(se_98_abundance, 'cohort', contigs, drop_zeros = T, contig_names = NULL)
```

<!-- ## Simulate variation in ANI with microbial load -->

<!-- Microbial load was predicted using https://github.com/grp-bork/microbial_load_predictor. -->
<!-- ```{r} -->
<!-- # Get the ids of cases after a random split -->
<!-- set.seed(789) -->
<!-- cascon = rbinom(round(SummarizedExperiment::ncol(data)),  1, 0.5) -->
<!-- cases = which(cascon == 1) -->

<!-- # setup metadata -->
<!-- meta_data = data.frame(run_accession = colnames(data), -->
<!--                        Case_status = cascon) -->


<!-- mic_load = data.frame(sequence = names(mic_load_orig), load = unname(mic_load_orig)) -->

<!-- mic_load$color = "green" -->

<!-- mic_load$color[mic_load$load < 4e10] = "blue" -->
<!-- mic_load$color[mic_load$load > 5.8e10] = "red" -->
<!-- plot(sort( log10(mic_load$load) ), col = sort(mic_load$color)) -->




<!-- data_orig <- read_sylph("TEST_DATA/zeevi/zeevi_PRJEB11532_query_gtdb_220_id98.tsv.gz") -->
<!-- data <- filter_by_presence(data_orig, min_nonzero = 50) -->
<!-- mic_load = readRDS("TEST_DATA/zeevi/microbial_load.rds") -->

<!-- # Get the ids of cases after a random split -->
<!-- set.seed(789) -->
<!-- cascon = rbinom(round(SummarizedExperiment::ncol(data)),  1, 0.5) -->
<!-- cases = which(cascon == 1) -->

<!-- # setup metadata -->
<!-- meta_data = data.frame(run_accession = colnames(data), -->
<!--                        Case_status = cascon) -->

<!-- meta_data$Case_status <- factor(meta_data$Case_status) -->
<!-- data = strainspy:::modify_metadata(se = data, meta_data = meta_data) -->

<!-- # This is somewhat arbitrary, but should work for simluations -->
<!-- mic_load$color = "green" -->
<!-- mic_load$color[mic_load$load < 6.25e10] = "blue" -->
<!-- mic_load$color[mic_load$load > 9e10] = "red" -->
<!-- plot(sort( log10(mic_load$load) ), col = sort(mic_load$color)) -->

<!-- meta_data$microbial_load_raw = mic_load$load[match(meta_data$run_accession, mic_load$sampleID)] -->

<!-- meta_data$microbial_load_HML = "M" -->
<!-- meta_data$microbial_load_HML[meta_data$microbial_load_raw > 9e10] = "H" -->
<!-- meta_data$microbial_load_HML[meta_data$microbial_load_raw < 6.25e10] = "L" -->
<!-- meta_data$microbial_load_HML = factor(meta_data$microbial_load_HML, levels = c("L", "M", "H")) -->


<!-- meta_data$microbial_load = (meta_data$microbial_load_raw - min(meta_data$microbial_load_raw))/ (max(meta_data$microbial_load_raw) - min(meta_data$microbial_load_raw)) -->

<!-- data = strainspy:::modify_metadata(se = data, meta_data = meta_data) -->

<!-- # # does the linear model work with this as well? -->
<!-- load_lm_fit = strainspy:::abundanceFit(data, as.formula('~microbial_load'), nthreads = 10) -->

<!-- # Use load as a predictor (numeric/normalised) -->
<!-- load_zib_fit = glmZiBFit(data,  as.formula('~microbial_load'), nthreads = 10) -->
<!-- load_ob_fit = glmFit(data,  as.formula('~microbial_load'), nthreads = 10) -->

<!-- # Keep high/medium/low 0/0.5/1 (factor) -->
<!-- loadHML_zib_fit = glmZiBFit(data,  as.formula('~microbial_load_HML'), nthreads = 10) -->
<!-- loadHML_ob_fit = glmFit(data,  as.formula('~microbial_load_HML'), nthreads = 10) -->

<!-- # Generate a violin plot for all hits -->
<!-- s1 = top_hits(load_zib_fit)$Contig_name -->
<!-- s2 = top_hits(load_ob_fit)$Contig_name -->
<!-- s3 = top_hits(loadHML_zib_fit)$Contig_name -->
<!-- s4 = top_hits(loadHML_ob_fit)$Contig_name -->

<!-- contig_names = c("Flavonifractor sp. DFI.6.63", "Lachnospiraceae bacterium HGM11888", -->
<!--                  "Lachnospiraceae bacterium BBMGS-G02-107", "Lachnospiraceae bacterium AM48-27BH AM48-27BH.Scaf1", -->
<!--                  "Lachnospiraceae bacterium AF58-1A AF58-1A.Scaf1", "Eubacteriales (uncult) ERR9762703", -->
<!--                  "Oscillibacter valericigenes DFI.6.76", "Oscillibacter sp. (uncult) ERR1600659", -->
<!--                  "Clostridiales bacterium HRGM_Genome_2357", "Lachnospiraceae (uncult) ERR1430444", -->
<!--                  "Lachnospiraceae bacterium AM48-27BH", "Bacteroides fragilis 1001271B", -->
<!--                  "Bacteroides sp. PHL 2737") -->

<!-- strainspy:::plot_violin(data, 'microbial_load_HML', unique(union(union(union(s1,s2), s3), s4)), contig_names, drop_ANI_zeros = T) -->

<!-- # Sylph abundance -->
<!-- data_abund = read_sylph("TEST_DATA/zeevi/zeevi_PRJEB11532_profile_gtdb_220_id98.tsv.gz", variable = "Taxonomic_abundance") -->
<!-- data_abund <- filter_by_presence(data_abund, min_nonzero = 50) -->
<!-- meta_data_sa = meta_data -->
<!-- meta_data_sa = meta_data_sa[sapply(colnames(data_abund), function(x) which(x == meta_data_sa$run_accession)),] -->

<!-- all(meta_data_sa$run_accession == colnames(data_abund)) -->

<!-- data_abund = strainspy:::modify_metadata(se = data_abund, meta_data = meta_data_sa) -->
<!-- load_lm_fit = strainspy::abundanceFit(data_abund, as.formula('~microbial_load'), nthreads = 10) -->


<!-- data_abund2 = data_abund[, which(data_abund$microbial_load_HML == "L" | data_abund$microbial_load_HML == "H")] -->
<!-- data_abund2 <- filter_by_presence(data_abund2, min_nonzero = 10) -->

<!-- data_abund2$microbial_load_HML = as.character(data_abund2$microbial_load_HML) -->
<!-- data_abund2$microbial_load_HML = factor(data_abund2$microbial_load_HML, labels = c("L", "H")) -->


<!-- top_hits(load_lm_fit) -->

<!-- strainspy:::plot_violin(data_abund, phenotype = "microbial_load_HML", top_hits(load_lm_fit)$Contig_name, contig_names = c("Mediterraneibacter butyricigenes", -->
<!--                                                                                                                           "Eubacterium segne", -->
<!--                                                                                                                           "Lachnospira hominis"), -->
<!--                         drop_ANI_zeros = T) -->

<!-- ### METAPHLAN IS DIFFICULT TO COMPARE, DROP THIS FOR NOW -->
<!-- # How bad is abundance in comparison? -->
<!-- # mp_files = dir("TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/", pattern = "*_profiled.tsv", full.names = T) -->
<!-- # strainspy::merge_metaphlan_files(metaphlan_files = mp_files, output_folder = "TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/") -->
<!-- # mp_sgb = strainspy::read_metaphlan("TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/metaphlan_merged.tsv.gz") -->
<!-- # -->
<!-- # # use the gtdb database for easier comparisons -->
<!-- # # merge_metaphlan_files(dir("TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/", "*gtdb.tsv", full.names = T), output_folder = "TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/", output_filename = "gtdb", sgb_database = F) -->
<!-- # mp_gtdb = strainspy::read_metaphlan("TEST_DATA/zeevi/mpa_vJun23_CHOCOPhlAnSGB_202307/gtdb_merged.tsv.gz") -->
<!-- # -->
<!-- # -->
<!-- # mp <- mp_gtdb -->
<!-- # mp <- filter_by_presence(mp, min_nonzero = 10) -->
<!-- # # meta_data needs to be renamed -->
<!-- # meta_data_mp = meta_data -->
<!-- # meta_data_mp$run_accession = paste(gsub("_1$", "", meta_data_mp$run_accession), "_profiled_gtdb", sep = "") -->
<!-- # meta_data_mp = meta_data_mp[unname(sapply(colnames(mp), function(x) which(x == meta_data_mp$run_accession))),] -->
<!-- # all(meta_data_mp$run_accession == colnames(mp)) # check -->
<!-- # -->
<!-- # mp = strainspy:::modify_metadata(se = mp, meta_data = meta_data_mp) -->
<!-- # -->
<!-- # load_lm_fit = strainspy:::abundanceFit(mp, as.formula('~microbial_load'), nthreads = 10) -->
<!-- # -->
<!-- # # Use load as a predictor (numeric/normalised) -->
<!-- # load_zib_mp_fit = glmZiBFit(mp,  as.formula('~microbial_load'), nthreads = 10) -->
<!-- # load_ob_mp_fit = glmFit(mp,  as.formula('~microbial_load'), nthreads = 10) -->
<!-- # -->
<!-- # # Keep high/medium/low 0/0.5/1 (factor) -->
<!-- # loadHML_zib_mp_fit = glmZiBFit(mp,  as.formula('~microbial_load_HML'), nthreads = 10) -->
<!-- # loadHML_ob_mp_fit = glmFit(mp,  as.formula('~microbial_load_HML'), nthreads = 10) -->
<!-- # -->
<!-- # -->
<!-- # s1 = top_hits(load_zib_mp_fit)$Contig_name -->
<!-- # s2 = top_hits(load_ob_mp_fit)$Contig_name -->
<!-- # s3 = top_hits(loadHML_zib_mp_fit)$Contig_name -->
<!-- # s4 = top_hits(loadHML_ob_mp_fit)$Contig_name -->
<!-- # -->
<!-- # # strainspy:::plot_violin(mp, 'microbial_load_HML', unique(union(union(union(s1,s2), s3), s4)), drop_ANI_zeros = T) -->
<!-- # strainspy:::plot_violin(mp, 'microbial_load_HML',s2, drop_ANI_zeros = T) -->

<!-- # Simulation steps: -->
<!-- # Vary identity of 1000 randomly selected features -->
<!-- # Ensure we are not picking cascon NAs -->
<!-- fit_cascon_null = readRDS("TEST_DATA/zeevi/TEMP_FITS/null_cascon.rds") -->
<!-- set.seed(789) -->
<!-- sel_1 = sample(which(!is.na(fit_cascon_null@coefficients$X.Intercept.)), 1000) -->

<!-- # we will vary the identity of these 1000 features as follows: -->
<!-- # Green isolates, scale_factor = 0.25 -->
<!-- # Green isolates, scale_factor = 0.25 -->
<!-- # Green isolates, scale_factor = 0.25 -->


<!-- # Randomly select 100 of these to vary with case/control status -->
<!-- set.seed(789) -->
<!-- sel = sample(sel_1, 100) -->
<!-- contigs = data@elementMetadata$Contig_name[sel] -->


<!-- ``` -->