---
title: "Untitled"
output: html_document
date: "2024-12-20"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Run Sylph v0.8.0

### UHGG

```
sylph query /data1/gerryt/sylph_analyses/uhgg_all_c200_v0.3.0.syldb -u --read-seq-id 99.9 -t 100 -1 ./*_1.fastq.gz -2 ./*_2.fastq.gz -o ./warren_sylph_query_uhgg_all_c200_v030.tsv

sylph profile /data1/gerryt/sylph_analyses/gtdb-r220-c200-dbv1.syldb -u --read-seq-id 99.9 -t 100 -1 ./*_1.fastq.gz -2 ./*_2.fastq.gz -o ./warren_sylph_profile_gtdb-r220-c200-dbv1.tsv
```

### GTDB 99%

```
sylph query /data1/gerryt/sylph_analyses/gtdb_99_ordered_DB/gtdb_ordered_99.syldb -u --read-seq-id 99.9 -t 100 -1 ./*_1.fastq.gz -2 ./*_2.fastq.gz -o ./wallen_sylph_query_gtdb_220_id99.tsv
```


## Clean up for examples

```{r}
library(tidyverse)
library(strainspy)
set.seed(789)

ena_meta <- read_tsv("~/Documents/wallen_example/filereport_read_run_PRJNA834801_tsv.txt")

meta <- read_csv("~/Documents/wallen_example/wallen_metadata.csv")
# meta <- meta |> add_column(run_accession = ena_meta$run_accession[match(meta$sample_name, ena_meta$library_name)], .before = 1)

keep_for_example <- meta |>
  group_by(Case_status) |>
  summarise(
    example_names = sample(run_accession, replace = FALSE, size = 100)
  )

sylph_profile <- read_tsv("~/Documents/wallen_example/wallen_sylph_profile_gtdb-r220-c200-dbv1.tsv.gz")
sylph_profile$Sample_file <- gsub("_1.*", "", basename(sylph_profile$Sample_file))


example_sylph_profile <- sylph_profile |> filter(Sample_file %in% keep_for_example$example_names)
tax <- read_taxonomy("./inst/extdata/sylph_gtdb_rep_taxonomy.tsv.gz")

example_sylph_profile <- example_sylph_profile[gsub("_genomic.*", "", gsub(".*/", "", example_sylph_profile$Genome_file)) %in% tax$Genome,]
example_meta <- meta |> filter(run_accession %in% keep_for_example$example_names)

write_tsv(example_sylph_profile, "~/Downloads/example_sylph_profile.tsv")
write_csv(example_meta, "~/Downloads/example_metadata.csv")
```

## Run full analysis



```{r}
library(SummarizedExperiment)
library(tidyverse)
library(strainspy)

meta_path <- "~/Documents/wallen_example/wallen_metadata.csv"
meta <- readr::read_csv(meta_path) 

# sylph_query_path <- "~/Documents/wallen_example/warren_sylph_query_uhgg_all_c200_v030.tsv.gz"
sylph_query_path <- "~/Documents/wallen_example/wallen_sylph_query_gtdb_220_id99.tsv.gz"
# sylph_query_path <- "~/Documents/wallen_example/wallen_sylph_profile_gtdb-r220-c200-dbv1.tsv.gz"

taxonomy <- read_taxonomy("~/Documents/wallen_example/bac_derep_ordered_d99_sylph_DB_taxonomy.tsv")

sy <- read_sylph(sylph_query_path)
sy <- filter_by_presence(sy, min_nonzero = 72)
colnames(sy) <- gsub("_1", "", colnames(sy))
dim(sy)

sy <- sy[colnames(sy) %in% meta$run_accession,]

colData(sy)$Sample_file <- gsub("_1", "", basename(colData(sy)$Sample_file))
meta$scale_read_count <- as.numeric(scale(meta$total_sequences))

temp <- merge(colData(sy), meta, by.x="Sample_file", by.y="run_accession", all.x=TRUE)
# Reorder the dataframe to match the original order
temp <- temp[match(colData(sy)$Sample_file, temp$Sample_file), ]
stopifnot(all(temp$Sample_file==colData(sy)$Sample_file))

# 490 cases and 234 controls (total 724)
table(temp$Case_status, useNA = "always")
#missing
temp$Sample_file[is.na(temp$Case_status)]

colData(sy) <- temp

sy <- sy[, !is.na(colData(sy)$collection_method)]
sy <- sy[, colData(sy)$collection_method!="swab"]


design <- as.formula("Case_status ~ Value + Age_at_collection + Sex + Do_you_drink_alcohol + Laxatives + Probiotic +
                     Pain_med + Depression_anxiety_mood_med + Antibiotics_current + Sleep_aid")

colData(sy)$Case_status <- factor(colData(sy)$Case_status)

system.time({
  logit_fit <- caseControlFit(sy,  design, nthreads = 4)
})

# saveRDS(logit_fit, "~/Documents/wallen_example/wallen_fit_profile_UHGG_logit.rds")
saveRDS(logit_fit, "~/Documents/wallen_example/wallen_logit_fit_profile_GTDB_99_logit.rds")

lth <- logit_fit@p_values
lth$adj.pvalue <- p.adjust(lth$Value, method="BH")
lth <- as_tibble(lth) %>% arrange(Value)

lth <- top_hits(logit_fit, coef = 2, alpha = 0.05, method="BH") %>% 
  arrange(p_value)

lth$Species <- taxonomy$Species[match(lth$Genome_file, taxonomy$Genome)]

lth %>% 
  group_by(Species) %>%
  summarise(min_pvalue= min(p_adjust))

table(lth$Species)

logit_fit@row_data

adj <- hadjust(logit_fit, taxonomy=taxonomy)

adj |> filter(p_adjust<0.05) %>% filter(Level=='Species')

plot_manhattan(logit_fit, taxonomy = taxonomy, method = 'HMP')

```


```{r}

design <- as.formula(" ~ Case_status + Age_at_collection + Sex + 
                     Do_you_drink_alcohol + Laxatives + Probiotic + Pain_med + 
                     Depression_anxiety_mood_med + Antibiotics_current + 
                     Sleep_aid")

system.time({
  beta_fit <- glmZiBFit(sy,  design, nthreads = 4)
})

saveRDS(beta_fit, "~/Documents/wallen_example/wallen_beta_fit_profile_GTDB_99.rds")

# beta_fit <- readRDS("~/Documents/wallen_example/wallen_beta_fit_profile_GTDB_99.rds")
 
th <- top_hits(beta_fit, coef = 2, alpha = 0.05, method="BH") %>%
  arrange(p_value) %>%
  filter(p_adjust < 0.05)
th[order(th$p_adjust),]

th$Species <- taxonomy$Species[match(th$Genome_file, taxonomy$Genome)]

# th %>% 
#   group_by(Species) %>%
#   summarise(min_pvalue= min(p_adjust))

table(th$Species)[grepl(".*coli.*", names(table(th$Species)))]

table(unique(th$Species) %in% unique(lth$Species))
table(unique(lth$Species) %in% unique(th$Species))


badj <- hadjust(beta_fit, taxonomy=taxonomy)

gg <- plot_manhattan(beta_fit, taxonomy = taxonomy, method = 'HMP')
gg + ggplot2::geom_point(aes(y = log_p_adjust, x = index_min + (index_max-index_min)/2), size=0.1) + 
  ggplot2::geom_segment(aes(y = log_p_adjust, x = index_min - 10, xend = index_max + 10), linewidth=3)


# cols <- c('#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928')
# 
# 
# pdf <- top_hits(beta_fit, coef = 2, alpha = 1, method="BH") 
# 
# pdf$Species <- taxonomy$Species[match(th$Genome_file, taxonomy$Genome)]
# nspecies <- length(unique(beta_fit@p_values$colour))
# 
# #Sylph paper pvalues plot
# ggplot(sylph_orig_pvalues, aes(y=-log10(p1), x=MAG, col=factor(colour))) +
#   geom_point(size=0.2) +
#   theme_bw() +
#   theme(legend.position = "none",
#         axis.text.x = element_blank(),
#         axis.ticks.x = element_blank()) +
#   scale_colour_manual(values = rep(cols, ceiling(nspecies/length(cols)))[1:nspecies])

```




<!-- Load pvalues form Sylph paper -->

<!-- ```{r} -->
<!-- library(tidyverse) -->

<!-- sylph_orig_pvalues <- read_tsv("~/Downloads/sylph-test-main/98-v05-covs-ordered_pvals.txt",  -->
<!--                                col_names = c("p1", "p2", "estimate", "padj", "colour", "MAG")) -->

<!-- sylph_orig_pvalues$MAG <- factor(sylph_orig_pvalues$MAG, levels = unique(sylph_orig_pvalues$MAG)) -->
<!-- sylph_orig_pvalues$colour <- factor(sylph_orig_pvalues$colour) -->

<!-- cols <- c('#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928') -->

<!-- nspecies <- length(unique(sylph_orig_pvalues$colour)) -->

<!-- #Sylph paper pvalues plot -->
<!-- ggplot(sylph_orig_pvalues, aes(y=-log10(p1), x=MAG, col=factor(colour))) + -->
<!--   geom_point(size=0.2) + -->
<!--   theme_bw() + -->
<!--   theme(legend.position = "none", -->
<!--         axis.text.x = element_blank(),  -->
<!--         axis.ticks.x = element_blank()) + -->
<!--   scale_colour_manual(values = rep(cols, ceiling(nspecies/length(cols)))[1:nspecies]) -->


<!-- logit_fit <- readRDS("~/Documents/wallen_example/wallen_fit_profile_UHGG_logit.rds") -->
<!-- lth <- top_hits(logit_fit, coef = 2, alpha = 1, method="BH") %>% arrange(p_value) -->
<!-- lth -->

<!-- merged_pdf <- sylph_orig_pvalues[,c("p1", "MAG", "colour")] %>% -->
<!--   filter(MAG %in% lth$Genome_file) %>% -->
<!--   rename(orig_sylph_pvalue = p1) -->

<!-- merged_pdf$new_p_value <- lth$p_value[match(merged_pdf$MAG, lth$Genome_file)] -->

<!-- # melt so that p1 and new_pvalue are in the same column -->
<!-- merged_pdf <- merged_pdf %>% -->
<!--   pivot_longer(cols = c(orig_sylph_pvalue, new_p_value),  -->
<!--                names_to = "method",  -->
<!--                values_to = "pvalue") -->

<!-- merged_pdf$MAG <- factor(as.character(merged_pdf$MAG), levels = unique(sylph_orig_pvalues$MAG)) -->

<!-- ggplot(merged_pdf, aes(y=-log10(pvalue), x=MAG, col=factor(colour))) + -->
<!--   geom_point(size=0.2) + -->
<!--   facet_wrap(~method, ncol=1) + -->
<!--   theme_bw() + -->
<!--   theme(legend.position = "none", -->
<!--         axis.text.x = element_blank(),  -->
<!--         axis.ticks.x = element_blank()) + -->
<!--   scale_colour_manual(values = rep(cols, ceiling(nspecies/length(cols)))[1:nspecies]) -->

<!-- ``` -->


<!-- ```{r} -->
<!-- testdf <- sylph_orig_pvalues[sample(1:nrow(sylph_orig_pvalues), size = 100, replace = FALSE), ]# %>% filter(-log10(p1) > 7) -->

<!-- rownames(sy) <- gsub("_1", "", rownames(sy)) -->
<!-- test_sy <- sy[rownames(sy) %in% testdf$MAG,] -->


<!-- which(rownames(test_sy)=="MGYG000157255") -->

<!-- #      run_to_covar[run] = [float(bases), float(age), sex, alc, lax, prob, pain, dep, anti, sleep] -->
<!-- design <- as.formula("Case_status ~ Value + Age_at_collection + Sex + Do_you_drink_alcohol + Laxatives + Probiotic + -->
<!--                      Pain_med + Depression_anxiety_mood_med + Antibiotics_current + Sleep_aid") -->
<!-- # design <- as.formula("Case_status ~ Value") -->

<!-- lapply(c("Age_at_collection", "Sex", "Do_you_drink_alcohol", "Laxatives", "Probiotic", -->
<!--   "Pain_med", "Depression_anxiety_mood_med", "Antibiotics_current", "Sleep_aid"), function(x) { -->
<!--     table(colData(test_sy)[,x]) -->
<!--   }) -->

<!-- colData(test_sy)$Case_status <- factor(colData(test_sy)$Case_status) -->
<!-- colData(test_sy)$scale_total_sequences <- scale(colData(test_sy)$total_sequences) -->

<!-- sort(colnames(colData(test_sy))) -->

<!-- temp_df <- colData(test_sy) -->
<!-- temp_df$Value <- SummarizedExperiment::assay(test_sy)[1,] -->

<!-- temp_df$Value <- temp_df$Value/100 -->
<!-- temp_df$Value[temp_df$Value<0.98] <- 0.98 -->

<!-- summary(glm(Case_status ~ Value, data = temp_df, family = binomial)) -->


<!-- system.time({ -->
<!--   test_fit <- caseControlFit(test_sy,  design, nthreads = 4)   -->
<!-- }) -->

<!-- test_hits <- as_tibble(test_fit@p_values) %>%  -->
<!--   add_column(MAG = rownames(test_sy), .before=1) %>% -->
<!--   arrange(Value)  -->


<!-- test_hits$sylph_orig <- testdf$p1[match(test_hits$MAG, testdf$MAG)] -->

<!-- test_hits[,c(1, 3,13)] -->

<!-- plot(test_hits[,c(3,13)]) -->
<!-- ``` -->
